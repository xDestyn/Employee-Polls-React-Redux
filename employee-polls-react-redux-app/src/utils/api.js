import {
  _getUsers,
  _getQuestions,
  _saveQuestion,
  _saveQuestionAnswer,
  formatQuestion,
} from './_DATA.js';

export const getInitialData = () => {
  return Promise.all([_getUsers(), _getQuestions()]).then(([users, questions]) => ({
    users,
    questions,
  }));
};

export const saveQuestion = (question) => {
  return _saveQuestion(question);
};

export const saveQuestionAnswer = (answer) => {
  return _saveQuestionAnswer(answer);
};

export const formatNewQuestion = (question) => {
  return formatQuestion(question);
};
