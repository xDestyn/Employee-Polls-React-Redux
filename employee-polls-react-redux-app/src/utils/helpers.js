/* eslint-disable no-empty-pattern */

import React from 'react';
import { useLocation, useNavigate, useParams } from 'react-router-dom';

export const formatDate = (timestamp) => {
  const date = new Date(timestamp);
  const time = date.toLocaleTimeString('en-US');
  return time.substr(0, 5) + time.slice(-2) + ' | ' + date.toLocaleDateString();
};

export const formatPoll = (question, author, authedUser) => {
  let hasAnswered = false;
  let isAuthor = false;
  let selectedVote = '';

  // Destructure poll
  const {
    id,
    optionOne: {
      votes: [],
      text: optionOneText,
    },
    optionTwo: {
      votes: [],
      text: optionTwoText,
    },
    timestamp,
  } = question;

  // Verify if a vote has been submitted
  // Extract that vote
  if (
    question.optionOne.votes.includes(authedUser) ||
    question.optionTwo.votes.includes(authedUser)
  ) {
    hasAnswered = true;
    if (question.optionOne.votes.includes(authedUser)) {
      selectedVote = question.optionOne.text;
    } else if (question.optionTwo.votes.includes(authedUser)) {
      selectedVote = question.optionTwo.text;
    }
  }

  // Verify if poll is self made
  if (author === authedUser) {
    isAuthor = true;
  }

  const optionOneVotes = question.optionOne.votes.length;
  const optionTwoVotes = question.optionTwo.votes.length;
  const totalVotes = optionOneVotes + optionTwoVotes;

  const optionOnePercentage = Math.round((optionOneVotes / totalVotes) * 100);
  const optionTwoPercentage = Math.round((optionTwoVotes / totalVotes) * 100);

  // Return Poll Model
  return {
    id,
    timestamp,
    name: author.name,
    avatarURL: author.avatarURL,
    optionOneText,
    optionTwoText,
    hasAnswered,
    selectedVote,
    isAuthor,
    totalVotes,
    optionOnePercentage,
    optionTwoPercentage,
  };
};

export const withRouter = (Component) => {
  const ComponentWithRouterProp = (props) => {
    let location = useLocation();
    let navigate = useNavigate();
    let params = useParams();
    return <Component {...props} router={{ location, navigate, params }} />;
  };

  return ComponentWithRouterProp;
};
