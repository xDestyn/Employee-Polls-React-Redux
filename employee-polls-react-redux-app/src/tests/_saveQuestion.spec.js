// @ts-nocheck
import { _saveQuestion } from '../utils/_DATA';

describe('_saveQuestion', () => {
  it('should return formatted object when correct arguments are received', async () => {
    const question = {
      optionOneText: 'A',
      optionTwoText: 'B',
      author: 'jsmith',
    };

    const response = await _saveQuestion(question);
    expect(response).toMatchSnapshot({
      author: 'jsmith',
      id: expect.any(String),
      optionOne: { text: 'A' },
      optionTwo: { text: 'B' },
      timestamp: expect.any(Number),
    });
  });

  it('should return error when incorrect arguments are received', async () => {
    const question = {
      optionOneText: undefined,
      optionTwoText: undefined,
      author: 'jsmith',
    };

    await expect(_saveQuestion(question)).rejects.toMatchSnapshot();
  });
});
