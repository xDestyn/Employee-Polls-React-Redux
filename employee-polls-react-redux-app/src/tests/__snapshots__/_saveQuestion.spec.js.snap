// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`_saveQuestion should return error when incorrect arguments are received 1`] = `"Please provide optionOneText, optionTwoText, and author"`;

exports[`_saveQuestion should return formatted object when correct arguments are received 1`] = `
Object {
  "author": "jsmith",
  "id": Any<String>,
  "optionOne": Object {
    "text": "A",
    "votes": Array [],
  },
  "optionTwo": Object {
    "text": "B",
    "votes": Array [],
  },
  "timestamp": Any<Number>,
}
`;
